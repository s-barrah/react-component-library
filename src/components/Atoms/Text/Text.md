
```jsx harmony
<div>
  <Text tag="h3" height="5">
    Heading 3 high
  </Text>
  <Text tag="h3" weight="1000">
    Heading 3 heavy
  </Text>
  <Text tag="h3" size="l">
    Heading 3 large
  </Text>
  <Text tag="p" colour="black">
    My paragraph
  </Text>
  <Text tag="p" size="xl" color="orange">
    My paragraph xl and orange
  </Text>
  <Text tag="p" size="xl" family="Anton">
    My paragraph xl and font Anton
  </Text>
</div>
```
