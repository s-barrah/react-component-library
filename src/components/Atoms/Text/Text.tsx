import * as React from 'react';
import * as PropTypes from 'prop-types';
import styled, { css } from 'styled-components';

interface ITheme {
  color?: (colorName: string) => string;
  fontSize?: (size: string) => string;
  fontFamily?: (family: string) => string;
  breakpoint?: (size: string) => string;
}

interface IBaseProps {
  family?: string;
  height?: string;
  weight?: string;
  color?: string;
  tag: string;
  size?: string;
  uppercase?: boolean;
  showLabel?: boolean;
  theme?: ITheme;
}

interface ITextProps extends IBaseProps {
  uppercase?: boolean;
}

/** Text renders different elements based on the `tag` prop
 *  Weight is checked for existence to prevent overriding the tag's css
 */
const BaseText: React.FC<IBaseProps> = ({
  tag,
  children,
  ...rest
}): JSX.Element => {
  const { uppercase, showLabel, ...validAttributes } = rest;
  return React.createElement(tag, validAttributes, children);
};

const Text = styled(BaseText)<ITextProps>`
  color: ${({ color, theme }) => (color ? theme?.color(color) : 'inherit')};
  font-size: ${({ size, theme }) => theme?.fontSize(size)};
  text-transform: ${({ uppercase }) => (uppercase ? 'uppercase' : 'inherit')};
  font-weight: ${({ weight }) =>
    weight ? `font-weight: ${weight}` : 'normal'};
  font-family: ${({ family, theme }) =>
    family ? theme?.fontFamily(family) : 'inherit'};
  ${({ size, theme }) =>
    size === 'super'
      ? css`
          font-size: ${theme?.fontSize('xxl')};
          line-height: 3rem;
          @media ${theme.breakpoint('small')} {
            font-size: ${theme?.fontSize('big')};
            line-height: ${theme?.fontSize('big')};
            margin-bottom: 1rem;
          }
          @media ${theme.breakpoint('medium')} {
            font-size: ${theme?.fontSize('super')};
            line-height: ${theme?.fontSize('super')};
            margin-bottom: 1rem;
          }
        `
      : null};
`;

Text.defaultProps = {
  family: 'Montserrat',
  weight: 'normal',
  height: undefined,
  uppercase: false,
  color: 'inherit',
  tag: 'span',
  size: 's',
};

Text.propTypes = {
  /** Font family */
  family: PropTypes.string,
  /** Weight of Font */
  weight: PropTypes.string,
  /** Line height */
  height: PropTypes.string,
  /** Sets text transform to uppercase. */
  uppercase: PropTypes.bool,
  /** Colors */
  color: PropTypes.string,
  /** Tag type */
  tag: PropTypes.string.isRequired,
  /** Sizes */
  size: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node,
    PropTypes.string,
  ]).isRequired,
};

export default Text;
