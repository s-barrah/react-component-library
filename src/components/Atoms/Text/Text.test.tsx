import * as React from 'react';
import 'jest-styled-components';

import renderWithTheme from '../../../hoc/shallowWithTheme';
import Text from './Text';

it('Atoms/Text component - renders xl yellow paragraph correctly', () => {
  const tree = renderWithTheme(
    <Text tag="p" size="xl" color="yellow">
      My paragraph xl and yellow
    </Text>
  ).toJSON();
  expect(tree).toMatchSnapshot();
});

it('Atom/Text - renders xl paragraph with Anton font correctly', () => {
  const component = renderWithTheme(
    <Text tag="p" size="xl" family="Anton">
      My paragraph xl and font Anton
    </Text>
  ).toJSON();
  expect(component).toMatchSnapshot();
});

it('Atoms/Text component - renders with props and text', () => {
  const props = {
    type: 'p',
    size: 'xl',
    family: 'Anton',
    weight: 'normal',
    color: 'inherit',
    text: 'My paragraph xl and font Anton',
  };
  const component = renderWithTheme(
    <Text tag={props.type} size={props.size} family={props.family}>
      {props.text}
    </Text>
  ).toJSON();
  expect(component?.props?.size).toBe(props.size);
  expect(component?.props?.family).toBe(props.family);
  expect(component?.props?.weight).toBe(props.weight);
  expect(component?.props?.color).toBe(props.color);
  expect(component?.children?.[0]).toBe(props.text);
});
