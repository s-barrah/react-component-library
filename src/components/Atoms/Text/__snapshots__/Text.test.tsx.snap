// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Atom/Text - renders xl paragraph with Anton font correctly 1`] = `
.c0 {
  font-size: 1.725rem;
  text-transform: inherit;
  font-weight: font-weight:normal;
  font-family: 'Anton',Helvetica,Arial,sans-serif;
}

<p
  className="c0"
  color="inherit"
  family="Anton"
  size="xl"
  weight="normal"
>
  My paragraph xl and font Anton
</p>
`;

exports[`Atoms/Text component - renders xl yellow paragraph correctly 1`] = `
.c0 {
  font-size: 1.725rem;
  text-transform: inherit;
  font-weight: font-weight:normal;
  font-family: 'Montserrat',Helvetica,Arial,sans-serif;
}

<p
  className="c0"
  color="yellow"
  family="Montserrat"
  size="xl"
  weight="normal"
>
  My paragraph xl and yellow
</p>
`;
